src/scheduler.cpp

```diff
@@ -0,0 +1,209 @@
+  *
+  */
+void SCH_initializeSchedule(DeploymentSchedule_t* pDeployment,
+                            system_tick_t startTime)
+{
+    uint32_t lastEndTime = 0;
+    while (pDeployment->init)
+    {
+        pDeployment->deploymentStartTime = startTime;
+        pDeployment->lastMeasurementTime = 0;
+        pDeployment->measurementCount = 0;
+        lastEndTime += pDeployment->maxDuration;
+
+        pDeployment->init(pDeployment);
+        pDeployment++;
+    }
+}
+
+
+
+/**
+ * @brief Retrieves the next scheduled event
+ *
+ * This function iterates through a schedule table to find the event that
+ * should run next based on the current system time.
+ * It ensures that no overlapping events are scheduled by checking with
+ * @ref SCH_willOverlap().
+ *
+ * @see tests/gtest.cpp for intended behavior
+ *
+ * @param scheduleTable Pointer to the first element of an array of
+ * DeploymentSchedule_t, which contains the scheduling information.
+ * @param p_nextEvent Pointer to a pointer where the next event to be executed
+ * will be stored.
+ * @param p_nextTime Pointer to where the time for the next event will
+ * be stored.
+ *
+ * @return Returns SUCCESS if a suitable event is found,
+ * TASK_SEARCH_FAIL otherwise.
+ */
+
+uint32_t SCH_getNextEvent(DeploymentSchedule_t* scheduleTable,
+                DeploymentSchedule_t** p_nextEvent, system_tick_t* p_nextTime)
+{
+
+    uint32_t minNextTime = UINT32_MAX;
+    DeploymentSchedule_t* nextEvent = nullptr;
+    system_tick_t currentTime = millis();
+
+
+    //! Iterate through each event in the schedule table.
+    for (int idx = 0; scheduleTable[idx].measure; ++idx)
+    {
+        //! Reference to current event in the schedule
+        DeploymentSchedule_t& currentEvent = scheduleTable[idx];
+        //! Reset the next run time for the current event
+        currentEvent.nextRunTime = UINT32_MAX;
+        //! Variable to store the calculated start time of the next event.
+        uint32_t nextStartTime;
+        //! Calculate first start time after delay.
+        uint32_t firstStartTime = currentEvent.deploymentStartTime +
+            currentEvent.ensembleDelay;
+
+        //! check if first event
+        if (currentTime <= firstStartTime)
+        {
+            if (firstStartTime < minNextTime &&
+                                            !SCH_willOverlap(scheduleTable,
+                                                idx, currentTime, firstStartTime))
+            {
+                minNextTime = firstStartTime;
+                nextEvent = &currentEvent;
+
+            }
+            continue;
+        }
+        //! Calculate intended count 
+        uint32_t intendedCount = (currentTime - firstStartTime - 1) /
+            currentEvent.ensembleInterval + 1;
+        uint32_t lastInterval = firstStartTime + (intendedCount - 1) *
+            currentEvent.ensembleInterval;
+        uint32_t nextInterval = lastInterval + currentEvent.ensembleInterval;
```

Comment:
Would it make more sense to simply retain a counter in the task state variables?